service/src/main/java/bio/terra/workspace/service/petserviceaccount/PetSaService.java

```diff
@@ -0,0 +1,258 @@
+package bio.terra.workspace.service.petserviceaccount;
+
+import bio.terra.cloudres.google.iam.ServiceAccountName;
+import bio.terra.workspace.common.exception.GcpException;
+import bio.terra.workspace.service.crl.CrlService;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.iam.SamRethrow;
+import bio.terra.workspace.service.iam.SamService;
+import bio.terra.workspace.service.petserviceaccount.model.UserWithPetSa;
+import bio.terra.workspace.service.workspace.GcpCloudContextService;
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import com.google.api.services.iam.v1.model.Binding;
+import com.google.api.services.iam.v1.model.Policy;
+import com.google.api.services.iam.v1.model.SetIamPolicyRequest;
+import com.google.common.collect.ImmutableList;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+import org.apache.http.HttpStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+/**
+ * This component holds logic for working with pet service accounts in a workspace. Pet Service
+ * accounts are a GCP-specific concept but are not resources as they are really managed by Sam.
+ * However, they exist in both Sam and GCP, and WSM often needs to interact with them.
+ */
+@Component
+public class PetSaService {
+
+  private static final Logger logger = LoggerFactory.getLogger(PetSaService.class);
+  private static final String SERVICE_ACCOUNT_USER_ROLE = "roles/iam.serviceAccountUser";
+
+  private final SamService samService;
+  private final GcpCloudContextService gcpCloudContextService;
+  private final CrlService crlService;
+
+  @Autowired
+  public PetSaService(
+      SamService samService, GcpCloudContextService gcpCloudContextService, CrlService crlService) {
+    this.samService = samService;
+    this.gcpCloudContextService = gcpCloudContextService;
+    this.crlService = crlService;
+  }
+
+  /**
+   * Wrapper around {@code enablePetServiceAccountImpersonationWithEtag} without requiring a
+   * particular GCP eTag value.
+   */
+  public Policy enablePetServiceAccountImpersonation(UUID workspaceId, UserWithPetSa userAndPet) {
+    // enablePetServiceAccountImpersonationWithEtag will only return an empty optional if the
+    // provided eTag does not match current policy. Because we do not use eTag checking here, this
+    // is always nonempty.
+    return enablePetServiceAccountImpersonationWithEtag(workspaceId, userAndPet, null).get();
+  }
+
+  /**
+   * Grant a user permission to impersonate their pet service account in a given workspace. Unlike
+   * other operations, this does not run as a flight because it only requires one write operation.
+   * This operation is idempotent.
```

Comment:
Sounds fine. Do we need retrying for this? Or just let the client retry on error?